### PR TITLE
Secure Video

### PR DESCRIPTION
**Pull request by california444**
**WORK IN PROGRESS**

This PR depends on the integration of:
- https://github.com/homebridge/HAP-NodeJS/pull/904
- https://github.com/homebridge/homebridge/pull/2973

Please understand this PR as a proof of concept or an initial version. I don't find enough time to make more progress , but I wanted to share the code as a lot of you have asked for it.
Big thanks to koush, I took over a lot of the "complex stuff" from his code base and tried to make it work with camera-ffmpeg and made more modifications and optimisations.
It is working more or less stable for me...

What is included:

- 2 new options in the settings under the experimental section:
-- recording: if enabled, the Secure Video Services are activated
-- prebuffer: if enabled the videostream from the camera gets continuously recorded and the prebuffer is filled 
- if motion and/or doorbell is enabled for a camera, the linked motion/doorbell services from Secure Video are reused
- motion/doorbell events trigger the handleFragments request of the new reocrdingDelegate
- fmp4 fragmentation according to the secure video spec
- prebuffering if activated

Open tasks:

- [ ] currently libx264 is hardcoded as codec. The video code should be configured according to the existing logic as for regular streaming
- [ ] audio is disabled as it caused some problems in my setup. 
- [ ] not sure how good error situations are handled. I guess it should be made a lot more robust
- [ ] I haven't tested how stable ffmpeg works if it runs "forever" for the prebuffering. Maybe it is a good idea to restart it once a while
- [ ] log outputs, especially of ffmpeg can definitely be improved
- [ ] maybe more options in the settings like configuring the prebuffer length
- [ ] ....

### CODE REVIEW

**Comment by Sunoo** [comment]
Fantastic. I'm happy to finish this up after the prerequisites are in place. Thanks so much for your efforts so far. I'll try to find time to really review this code this weekend, then I'll probably pull it into a branch here.

**Comment by Stetocasa** [comment]
@Sunoo any potential beta release date to test HKSV on your Camera plugin?

**Comment by Sunoo** [comment]
@Stetocasa It should be shortly after HAP-NodeJS and Homebridge merge in support.

**Comment by Stetocasa** [comment]
Thanks @Sunoo do you know which issue follow on GitHub to know on time for the merge release ?

**Comment by Sunoo** [comment]
@Stetocasa This PR depends on https://github.com/homebridge/HAP-NodeJS/pull/904 and https://github.com/homebridge/homebridge/pull/2973

**Comment by california444** [comment]
Actually there is a third, optional PR directly against the HAP-NodeJS fork from koush to add doorbell support
[https://github.com/koush/HAP-NodeJS/pull/1](https://github.com/koush/HAP-NodeJS/pull/1).
It seems like due to missing rights I can not add commit to the PR in HAP-NodeJS

**Comment by Sunoo** [comment]
@california444 Ah, wasn’t aware of that one, thanks for pointing that out. If that doesn’t make it in the next Homebridge release for whatever reason, I’d probably merge support for motion to this plugin, then add doorbell later whenever that gets pulled in.

**Comment by bdruth** [comment]
Also, quick-followup, when I look closer at the errors, they seem odd for just being issues around not seeing the right dependent code, e.g.:
```
src/streamingDelegate.ts:198:24 - error TS2565: Property 'recordingDelegate' is used before being assigned.

198         delegate: this.recordingDelegate
```

**Comment by california444** [comment]
@bdruth in tsconfig.json try setting strict to false. I remember I changed something to make progress faster ;-) maybe I did not turn it back before committing…

**Comment by california444** [comment]
> @california444 I may be misunderstanding the `prebuffer` approach, but is the gist of that, that `camera-ffmpeg` is always streaming from the camera, effectively? If so, why does the plugin launch a new `ffmpeg` process when opening the camera stream in Home? Would it be possible to just shunt whatever stream is being used for `prebuffer`?

If prebuffer is enabled the camera video is streamed all the time in order to fill the prebuffer so that we have the last 4 seconds of an motion/doorbell event in the buffer. 

Please correct me if I’m wrong, but I guess for each open camera stream in HOME the plug-in starts a new ffmpeg process in the current implementation. 
If I remember correct ffmpeg can handle multiple output streams. So in general I think your approach could work, but would require a larger rework to support more than one video stream. And for a regular stream you don‘t want to have the prebuffer. So some more logic to differentiate between the streams….

**Comment by SeydX** [comment]
Hey @california444 

I used your HSV implementation for my ``homebridge-camera-ui`` plugin and modified it slightly (to make it compatible with the user interface)

https://github.com/SeydX/homebridge-camera-ui/tree/hsv

HSV works ok so far with no errors, only FFmpeg doesn't always cooperate well. When I manually trigger a movement, sometimes the following error message comes up: `Error writing trailer of tcp://127.0.0.1:34837: Broken pipe`.

Due to the error, the recording is not saved with HSV (missing in the timeline) but the recording exists (because it is also passed to the user interface and processed).

Have you had this error message?

Also, I think the prebuffering is awesome, thanks for that.

A small note: If the connection to the camera is interrupted and re-established, it would be great if the prebuffering is restarted or?

**Comment by california444** [comment]
@SeydX good to hear it is working in  general. I did not experience this problem.  When does this error occur. Right at the beginning of the stream? Maybe the debug logs provide more information?

**Comment by SeydX** [comment]
@california444 

I have not been able to debug it any further. Error occurs only at the end when recording is finished

**Comment by california444** [comment]
@SeydX did you change the prebuffer length? There is a 30s timeout in one of the servers. If the prebuffer +fragments exceed 30s that could maybe an issue…

**Comment by arkadicolson** [comment]
I would like to test this to. What's the easiest way to get the code on my homebridge?

**Comment by california444** [comment]
@Sunoo There is a new [HAP-NodeJS PR 920](https://github.com/homebridge/HAP-NodeJS/pull/920), which will hopefully bring HSV support soon. As far as I can see the API has changed compared to the initial PR 904, so that this PR here will no longer work and needs to be modified.
Are you going to take over this PR? I probably will not find enough time in the near future to make it work with acceptable code quality....

**Comment by Sunoo** [comment]
@california444 Yea, I’ve already been looking into that new PR. Now that it seems support is actually coming, I’m happy to start working on this in earnest. Hopefully will put a bunch of work in today, but we’ll see how my day job goes.

**Comment by adx74** [comment]
Hi, [HAP-NodeJS PR 920](https://github.com/homebridge/HAP-NodeJS/pull/920) has been merged.

That’s mean that we can now have HSV ? 
And thanks @Sunoo for this work !

**Comment by Sunoo** [comment]
@adx74 My plan is to get at least initial support in place on my plugin this weekend. Homebridge will still need to also merge that PR before it is easily usable by everyone though.

**Comment by Sunoo** [comment]
I would expect it not to take too long, most of the changes were in HAP.

**Comment by california444** [comment]
With the changed API maybe more exports are now required in the homebridge PR. 
@Sunoo I am not at home right now. If you find missing exports feel free to add them to the homebridge PR. Otherwise I try to look at it as soon as possible.

**Comment by Sunoo** [comment]
@california444 I’m definitely gonna take over this PR. I’ll try to take a look at the HB PR after my day job today, my recollection is that it was pretty slim though.

File: config.schema.json [hunk 1]
```diff
@@ -247,6 +247,20 @@
                 "title": "Two-way FFmpeg Debug Logging",
                 "type": "boolean",
                 "description": "Includes debugging output from the FFmpeg process used for two-way audio in the Homebridge log."
+              },
+              "recording": {
+                "title": "Enable Homekit Secure Video Recording",
+                "type": "boolean",
+                "required": true,
+                "default": false,
+                "description": "Enables Homekit Secure Video and records video on motion and doorbell events. Requires a home hub and iCloud plan with at least 200GB of storage."
```
Comment (nanosonde) [inline config.schema.json:256]: Is this comment still true? IMHO, Apple changed this. The 50GB plan also offers support for a Homekit Secure Video Camera. However, only one is supported.
See https://support.apple.com/en-us/HT201238

File: src/recordingDelegate.ts [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            let input:string[] = await this.preBuffer.getVideo(configuration.mediaContainerConfiguration.prebufferLength);
+            ffmpegInput.push(...input);
+        }
+        else {
+            ffmpegInput.push(...this.videoConfig.source.split(" "));
+        }
+        
+        this.log.debug("Start recording...", this.cameraName);
+        
+        const session = await this.startFFMPegFragmetedMP4Session(this.videoProcessor, ffmpegInput, audioArgs, videoArgs);
```
Comment (nanosonde) [inline src/recordingDelegate.ts:180]: Typo in startFFMPegFragmetedMP4Session

**Comment by schwie** [comment]
Presumably the team is already aware of this, if not it could be helpful:

[https://github.com/Supereg/secure-video-specification](https://github.com/Supereg/secure-video-specification)

**Comment by albinmedoc** [comment]
What happend with this PR. This feature doesn't exist in the hksv branch... even if it has been merged into it.

**Pull request merged**